File: desktop-tauri/src/ipc/account.ts
import { invoke } from "@tauri-apps/api/core";
import { listen, type UnlistenFn } from "@tauri-apps/api/event";

import type { AccountSnapshot } from "../types";

export function accountRestore(): Promise<AccountSnapshot> {
  return invoke("account_restore");
}

export function accountLoginEmail(email: string, password: string): Promise<AccountSnapshot> {
  return invoke("account_login_email", { email, password });
}

export function accountSignupEmail(
  name: string,
  email: string,
  password: string,
): Promise<AccountSnapshot> {
  return invoke("account_signup_email", { name, email, password });
}

/** The code half of a login. The challenge id is held natively. */
export function accountTwoFactorSubmit(code: string): Promise<AccountSnapshot> {
  return invoke("account_two_factor_submit", { code });
}

export function accountTwoFactorCancel(): Promise<AccountSnapshot> {
  return invoke("account_two_factor_cancel");
}

export function accountOauthStart(provider: string): Promise<AccountSnapshot> {
  return invoke("account_oauth_start", { provider });
}

export function accountOauthCancel(): Promise<AccountSnapshot> {
  return invoke("account_oauth_cancel");
}

export function accountProfileRefresh(): Promise<AccountSnapshot> {
  return invoke("account_profile_refresh");
}

export function accountProfileUpdate(name: string, email: string): Promise<AccountSnapshot> {
  return invoke("account_profile_update", { name, email });
}

export function accountPasswordForgot(email: string): Promise<string> {
  return invoke("account_password_forgot", { email });
}

export function accountResendVerification(): Promise<string> {
  return invoke("account_resend_verification");
}

/** The account photo as a `data:` URL; null when there is none. Fetched
 * natively because the renderer's content policy cannot reach the network. */
export function accountAvatar(): Promise<string | null> {
  return invoke("account_avatar");
}

export function accountLogout(): Promise<AccountSnapshot> {
  return invoke("account_logout");
}

export function accountOpenLegal(page: "privacy" | "terms"): Promise<void> {
  return invoke("account_open_legal", { page });
}

export function onAccountChanged(
  callback: (snapshot: AccountSnapshot) => void,
): Promise<UnlistenFn> {
  return listen<AccountSnapshot>("account:changed", (event) => callback(event.payload));
}

File: desktop-tauri/src-tauri/src/commands/account.rs
use tauri::{AppHandle, State};

use crate::account_auth;
use crate::account_oauth;
use crate::account_profile;
use crate::account_types::AccountSnapshot;
use crate::state::AppState;

#[tauri::command]
pub fn account_snapshot(state: State<'_, AppState>) -> AccountSnapshot {
    state.account.snapshot()
}

#[tauri::command]
pub async fn account_restore(state: State<'_, AppState>) -> Result<AccountSnapshot, String> {
    Ok(account_auth::restore(&state).await)
}

#[tauri::command]
pub async fn account_login_email(
    state: State<'_, AppState>,
    email: String,
    password: String,
) -> Result<AccountSnapshot, String> {
    Ok(account_auth::login_email(&state, email, password).await)
}

#[tauri::command]
pub async fn account_signup_email(
    state: State<'_, AppState>,
    name: String,
    email: String,
    password: String,
) -> Result<AccountSnapshot, String> {
    Ok(account_auth::signup_email(&state, name, email, password).await)
}

#[tauri::command]
pub async fn account_oauth_start(
    app: AppHandle,
    provider: String,
) -> Result<AccountSnapshot, String> {
    Ok(account_oauth::start(app, provider).await)
}

#[tauri::command]
pub fn account_oauth_cancel(state: State<'_, AppState>) -> AccountSnapshot {
    state.account.cancel_oauth();
    state.account.snapshot()
}

#[tauri::command]
pub async fn account_profile_refresh(
    state: State<'_, AppState>,
) -> Result<AccountSnapshot, String> {
    account_profile::refresh(&state).await
}

#[tauri::command]
pub async fn account_profile_update(
    state: State<'_, AppState>,
    name: String,
    email: String,
) -> Result<AccountSnapshot, String> {
    account_profile::update(&state, name, email).await
}

#[tauri::command]
pub async fn account_password_forgot(email: String) -> Result<String, String> {
    account_profile::password_forgot(email).await
}

#[tauri::command]
pub async fn account_resend_verification(state: State<'_, AppState>) -> Result<String, String> {
    account_profile::resend_verification(&state).await
}

#[tauri::command]
pub async fn account_logout(state: State<'_, AppState>) -> Result<AccountSnapshot, String> {
    Ok(account_auth::logout(&state).await)
}

/// Opens one of the enumerated Vibyra legal pages in the system browser.
/// The renderer names a page, never a URL.
#[tauri::command]
pub fn account_open_legal(page: String) -> Result<(), String> {
    let url = match page.as_str() {
        "privacy" => "https://vibyra.app/legal/privacy",
        "terms" => "https://vibyra.app/legal/terms",
        _ => return Err("Unknown legal page.".to_owned()),
    };
    crate::provider_auth_url::open(url)
}
